In [1]:
# import libraries
import pandas as pd
import numpy as np
import os 

In [3]:
path = r"C:\Users\siram\OneDrive\Desktop\Instacart Basket Analysis"

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3_orders_products','orders.csv'), index_col = False)

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3_orders_products','products.csv'), index_col = False)

# consistency checks of df_prods dataframe

In [17]:
# looking for mixed data type in df_prods
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)
    #the column product_name have mixed data type

product_name


In [18]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# fixing product_name mixed data type

In [26]:
df_prods['product_name'] = df_prods['product_name'].astype('str')


In [20]:
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)
#Now all the data into the df_prods columns has the same data type

# looking for missing values df_prods

In [22]:
df_prods.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [30]:
df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [24]:
df_nan

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [29]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## 1. The product_id max number doesn't fit with the count number, so 5 products have the same product_id or don't have any product_id.
## 2. The max price looks to high if we compare it with the mean and the %.


In [7]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [10]:
#create a data frame
df_test = pd.DataFrame()

In [11]:
df_test['mix']=['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

# looking for mixed data type in df_ords

In [15]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
    #no mixed data

# looking for missing values in df_ords

In [31]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## we find 206209 missing values in the column day_since_prior_order. My interpretation of this ammount of missing values is that probably this users are new, so they never bought anything before.

In [35]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [56]:
new_customer = []

In [58]:
new_customer = []
for row in df_ords['days_since_prior_order']:
    if row > 0 : new_customer.append('old_customer')
    else : new_customer.append('new_customer')

In [59]:
df_ords['new_customer'] = new_customer 

In [61]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  new_customer  
0                     NaN  new_customer  
1                    15.0  old_customer  
2                    21.0  old_customer  
3                    29.0  old_customer  
4                    28.0  old_customer

## I choose to create a new column to expecificate is the customer are new or old, because eliminating the nules will means to eliminate all the new customers. 

# check for duplicate values in your df_ords data

In [63]:
df_dups=df_ords[df_ords.duplicated()]

In [64]:
df_dups
#there are no duplicates

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

# Exporting data

In [65]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [66]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))